GitUI/BranchTreePanel/RepoObjectsTree.Nodes.cs

```diff
@@ -182,34 +190,37 @@ protected void ReloadNodes(Func<CancellationToken, Task> loadNodesTask)
                         token.ThrowIfCancellationRequested();
                         await TreeViewNode.TreeView.SwitchToMainThreadAsync();
 
-                        FillTreeViewNode();
+                        FillTreeViewNode(token, _firstReloadNodesSinceModuleChanged);
                     }
                     finally
                     {
-                        IgnoreSelectionChangedEvent = false;
-                        TreeViewNode.TreeView.EndUpdate();
-                        repoObjectTree.Enabled = true;
-                        ExpandPathToSelectedNode();
+                        // Avoid accessing disposed objects if we're reloading nodes while GE is closing
```

Comment:
It needs a little rework. The mess was introduced in c9e47e4b3e79792b1afcc762e0164445c82606a6 
This finally block is always executed, regardless of the token being cancelled or not. If the token is cancelled then there is a race condition on the boolean variables). BeginUpdate/EndUpdate spanning async operations seems to wrong in my opinion. I would avoid it at any cost. It is not guaranteed that the finally block executes on the main thread.
`_reloadCancellationTokenSequence` needs to be disposed. The whole logic needs to be rethought again.
I will try to find a few hours this weekend to send a PR that will fix this.

Reply:
Thanks @jbialobr.

> This finally block is always executed, regardless of the token being cancelled or not. If the token is cancelled then there is a race condition on the boolean variables).

I see what you're saying about the race condition on Enabled, and node selection. I can add back the check for token cancellation as it was before. 

> BeginUpdate/EndUpdate spanning async operations seems to be wrong in my opinion. I would avoid it at any cost.

This was there before https://github.com/gitextensions/gitextensions/commit/c9e47e4b3e79792b1afcc762e0164445c82606a6, though. Are you saying that this was a bad idea even before that change?

> I will try to find a few hours this weekend to send a PR that will fix this.

Would you be up to fixing it on this branch? Or will you fix it on master? With some guidance, I can do these fixes as well.

Reply:
>This was there before [c9e47e4](https://github.com/gitextensions/gitextensions/commit/c9e47e4b3e79792b1afcc762e0164445c82606a6), though. Are you saying that this was a bad idea even before that change?

Before this method looked like this:

```csharp
 public async Task ReloadAsync(CancellationToken token)
            {
                await TreeViewNode.TreeView.SwitchToMainThreadAsync(token);
                ClearNodes();

                await LoadNodesAsync(token).ConfigureAwait(false);

                await TreeViewNode.TreeView.SwitchToMainThreadAsync(token);
                TreeViewNode.TreeView.BeginUpdate();
                try
                {
                    FillTreeViewNode();
                    if (TreeViewNode.TreeView.SelectedNode != null)
                    {
                        TreeViewNode.TreeView.SelectedNode.EnsureVisible();
                    }
                    else if (TreeViewNode.TreeView.Nodes.Count > 0)
                    {
                        TreeViewNode.TreeView.Nodes[0].EnsureVisible();
                    }
                }
                finally
                {
                    TreeViewNode.TreeView.EndUpdate();
                }
            }
```
And there is no async operation between BeginUpdate and EndUpdate or I am missing sth.

>Would you be up to fixing it on this branch? Or will you fix it on master?

I will send a PR against the master branch.

>With some guidance, I can do these fixes as well.

I prefer to write the code myself, because seeing the corrected version it is easier to evaluate if I am not missing sth and if all works well.

Reply:
> And there is no async operation between BeginUpdate and EndUpdate or I am missing sth.

You're right. The reason I changed this code was because I needed each of the root nodes (Branches, Remotes, Tags, and now Submodules) to be able to update independently. This is necessary, especially for the Submodules node, so that it can update only when submodule changes happen, via a pub/sub system. Because of this, each root node can start up an async update, and each will do the Begin/EndUpdate, as you pointed out.

I think we can modify this function to behave similarly to the old code such that the async operation happens first, and outside the try/catch, and we do the UI update afterwards. Something like this (I haven't tested it yet):

```csharp
            protected void ReloadNodes(Func<CancellationToken, Task> loadNodesTask)
            {
                ThreadHelper.ThrowIfNotOnUIThread();

                ThreadHelper.JoinableTaskFactory.RunAsync(async () =>
                {
                    Nodes.Clear();

                    var token = _reloadCancellationTokenSequence.Next();
                    await loadNodesTask(token);
                    await TreeViewNode.TreeView.SwitchToMainThreadAsync();
                    try
                    {
                        TreeViewNode.TreeView.BeginUpdate();
                        TreeViewNode.TreeView.Enabled = false;
                        IgnoreSelectionChangedEvent = true;
                        FillTreeViewNode();
                    }
                    finally
                    {
                        IgnoreSelectionChangedEvent = false;
                        TreeViewNode.TreeView.Enabled = true;
                        TreeViewNode.TreeView.EndUpdate();
                        ExpandPathToSelectedNode();
                    }
                }).FileAndForget();
            }
```

Reply:
I tried this change, and it seems to be working well. I'll do some more testing, and update the branch. @jbialobr perhaps you won't need to revisit this after all...

Reply:
Okay, I removed the commit that added the Dipose checks, and pushed a new commit with the changes described above. @jbialobr, let me know if this is satisfactory. Cheers.